app/models/manageiq/providers/vmware/network_manager/refresh_parser.rb

```diff
@@ -158,20 +227,99 @@ def subnet_name(network)
       "subnet-#{network.name}"
     end
 
-    def vapp_network_id(name, vapp)
-      "#{vapp.id}_#{name}"
-    end
-
     def vapp_network_name(name, vapp)
       "#{name} (#{vapp.name})"
     end
 
-    def port_id(net_if)
-      "#{net_if[:vm].id}#NIC##{net_if[:idx]}"
+    def port_id(nic_data)
+      "#{nic_data[:vm].ems_ref}#NIC##{nic_data[:NetworkConnectionIndex]}"
     end
 
-    def port_name(net_if)
-      "#{net_if[:vm].name}#NIC##{net_if[:idx]}"
+    def port_name(nic_data)
+      "#{nic_data[:vm].name}#NIC##{nic_data[:NetworkConnectionIndex]}"
+    end
+
+    def floating_ip_id(nic_data)
+      "floating_ip-#{port_id(nic_data)}"
+    end
+
+    # vCD API does not provide us with vApp network IDs for some reason. Luckily it provides
+    # "Links" section whith API link to edit network page and network ID is part of this link.
+    def network_id_from_links(data)
+      return unless data[:Link]
+      links = to_list(data[:Link])
+      links.each do |link|
+        m = /.*\/network\/(?<id>[^\/]+)\/.*/.match(link[:href])
+        return m[:id] unless m.nil? || m[:id].nil?
+      end
+      nil
+    end
+
+    # Detect when network configuration as reported by vapp is actually a VDC network.
+    # In such cases vCD reports duplicate of VDC networks (all the same, only ID is different)
+    # instead the original one, which would result in duplicate entries in the VMDB. When the
+    # function above returns not nil, such network was detected. The returned value is then the
+    # actual VDC network specification.
+    def corresponding_vdc_network(net_conf, vdc_networks)
+      if net_conf.dig(:networkName) == net_conf.dig(:Configuration, :ParentNetwork, :name)
+        parent_vdc_network(net_conf, vdc_networks)
+      end
+    end
+
+    def parent_vdc_network(net_conf, vdc_networks)
+      vdc_networks[net_conf.dig(:Configuration, :ParentNetwork, :id)]
+    end
+
+    # Remember network id for given network name. Generally network names are not unique,
+    # but inside vapp network specification they are. Therefore we must remember what network
+    # id was listed for given network name in corresponding vapp in order to be able to later
+    # hook VM to the appropriate network (VM only reports network name, without network ID...).
+    def memorize_network_name_mapping(vapp_id, network_name, network_id)
+      @network_name_mapping[vapp_id] ||= {}
+      @network_name_mapping[vapp_id][network_name] = network_id
+    end
+
+    def read_network_name_mapping(vapp_id, network_name)
+      @network_name_mapping.dig(vapp_id, network_name)
+    end
+
+    def to_cidr(address, netmask)
+      return unless address.to_s =~ Resolv::IPv4::Regex && netmask.to_s =~ Resolv::IPv4::Regex
+      address + '/' + netmask.to_s.split(".").map { |e| e.to_i.to_s(2).rjust(8, "0") }.join.count("1").to_s
+    end
+
+    def to_list(list_or_hash)
+      list_or_hash.kind_of?(Array) ? list_or_hash : [list_or_hash].compact
```

Comment:
Should be able to just use `Array(abcd)` or `abcd.to_a` everywhere instead of this method

Reply:
Unfortunately it ain't that simple because of this one example where `Array()` behaves really strange :

```ruby
Array({ :key => 'val'})
=> [[:key, "val"]]  # we expect [{ :key => 'val'}]
```

Reply:
🙄 you're right that is weird, looks like `Array.wrap` does the right thing though

Reply:
You're right about `Array.wrap`, it works like a charm. Thanks, I've repushed.